src/EditorFeatures/Core/EditorConfigSettings/Updater/SettingsUpdateHelper.cs

```diff
@@ -214,87 +260,126 @@ internal static partial class SettingsUpdateHelper
-            {
-                var groups = s_headerPattern.Match(curLineText.Trim()).Groups;
-                mostRecentHeaderText = groups[1].Value.ToString().ToLowerInvariant();
-                return mostRecentHeaderText.Equals("*", StringComparison.Ordinal);
-            }
+            TextLine? lastValidSpecificHeader = null;
+            TextLine? lastValidSpecificHeaderSpanEnd = null;
 
-            static (string fileName, string[] splicedFileExtensions) ParseHeaderParts(string mostRecentHeaderText)
+            var textChange = new TextChange();
+            foreach (var curLine in editorConfigText.Lines)
             {
-                // We splice on the last occurrence of '.' to account for filenames containing periods.
-                var nameExtensionSplitIndex = mostRecentHeaderText.LastIndexOf('.');
-                var fileName = mostRecentHeaderText.Substring(0, nameExtensionSplitIndex);
-                var splicedFileExtensions = mostRecentHeaderText[(nameExtensionSplitIndex + 1)..].Split(',', ' ', '{', '}');
-
-                // Replacing characters in the header with the regex equivalent.
-                fileName = fileName.Replace(".", @"\.");
-                fileName = fileName.Replace("*", ".*");
-                fileName = fileName.Replace("/", @"\/");
-
-                return (fileName, splicedFileExtensions);
-            }
+                var curLineText = curLine.ToString();
+                if (IsOption(curLineText))
+                {
+                    var (untrimmedKey, key, value, severity, comment) = GetOptionParts(curLineText);
 
-            static bool IsNotEmptyOrComment(string currentLineText)
-            {
-                return !string.IsNullOrWhiteSpace(currentLineText) && !currentLineText.Trim().StartsWith("#", StringComparison.OrdinalIgnoreCase);
-            }
+                    // Verify the most recent header is a valid header
+                    if (IsValidHeader(mostRecentHeader, lastValidHeader) &&
+                        string.Equals(key, optionName, StringComparison.OrdinalIgnoreCase))
+                    {
+                        // We found the rule in the file -- replace it with updated option value.
+                        textChange = new TextChange(curLine.Span, $"{untrimmedKey}={optionValue}{comment}");
+                    }
+                }
+                else if (IsHeader(curLineText))
+                {
+                    mostRecentHeader = curLine;
+                    if (ShouldSetAsLastValidHeader(curLineText, out var mostRecentHeaderText))
+                    {
+                        lastValidHeader = mostRecentHeader;
+                    }
+                    else
+                    {
+                        var (fileName, splicedFileExtensions) = GetHeaderParts(mostRecentHeaderText);
+                        if ((relativePath.IsEmpty() || new Regex(fileName).IsMatch(relativePath)) &&
```

Comment:
you can pull out the `new Regex` pieces to statics.  they're stateless and can be reused.